File: boj/boj_7562.cpp
#include <iostream>
#include <queue>
#include <memory.h>

int main()
{
	int dx[8] = { -2, -2, -1, -1,  1, 1, 2, 2 };
	int dy[8] = { -1, 1, -2, 2, -2, 2, -1, 1 };
	bool visit[301][301];
	int t = 0;
	std::cin >> t;

	while (t--)
	{
		std::queue<std::pair<std::pair<int, int>, int>> q;
		int startX, startY, endX, endY, size;
		std::cin >> size;
		std::cin >> startX >> startY >> endX >> endY;

		memset(visit, false, sizeof visit);
		q.push(std::make_pair(std::make_pair(startX, startY), 0));
		visit[startY][startX] = true;

		while (!q.empty())
		{
			std::pair<int, int> cur = q.front().first;
			int curX = cur.first;
			int curY = cur.second;
			int step = q.front().second;
			q.pop();

			if (curX == endX && curY == endY)
			{
				std::cout << step << "\n";
				break;
			}

			for (int i = 0; i < 8; ++i)
			{
				int nextX = curX + dx[i];
				int nextY = curY + dy[i];

				if (nextX >= 0 && nextX < size && nextY >= 0 && nextY < size && !visit[nextY][nextX])
				{
					q.push(std::make_pair(std::make_pair(nextX, nextY), step + 1));
					visit[nextY][nextX] = true;
				}
			}
		}
	}
	return 0;
}

File: boj/boj_7562.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

class Move {
	public int x;
	public int y;
	public int step;

	public Move(int x, int y, int step) {
		this.x = x;
		this.y = y;
		this.step = step;
	}
}

public class Main {
	public static int dx[] = { -2, -2, -1, -1, 1, 1, 2, 2 };
	public static int dy[] = { -1, 1, -2, 2, -2, 2, -1, 1 };
	public static boolean visit[][];

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);

		int t = sc.nextInt();
		Queue<Move> q = new LinkedList<Move>();
		while (t-- > 0) {
			q.clear();
			int n = sc.nextInt();
			int startX = sc.nextInt();
			int startY = sc.nextInt();
			int endX = sc.nextInt();
			int endY = sc.nextInt();

			visit = new boolean[n][n];
			q.add(new Move(startX, startY, 0));
			visit[startY][startX] = true;
			while (!q.isEmpty()) {
				int curX = q.peek().x;
				int curY = q.peek().y;
				int step = q.peek().step;
				q.poll();
				
				if (curX == endX && curY == endY) {
					System.out.println(step);
					break;
				}

				for (int i = 0; i < 8; ++i) {
					int nextX = curX + dx[i];
					int nextY = curY + dy[i];
					if (nextX >= 0 && nextX < n && nextY >= 0 && nextY < n && !visit[nextY][nextX]) {
						q.add(new Move(nextX, nextY, step + 1));
						visit[nextY][nextX] = true;
					}
				}
			}

		}

	}
}
